src/rust/ensogl/src/display/style.rs

```diff
@@ -0,0 +1,686 @@
+// === Expression ===
+// ==================
+
+/// Expression of a style sheet.
+#[derive(Derivative)]
+#[derivative(Debug)]
+pub struct Expression {
+    /// Indexes of all vars which are used as sources to the function of this expression.
+    pub sources : Vec<Index<Var>>,
+    /// Function used to compute the new value of the style sheet.
+    #[derivative(Debug="ignore")]
+    function : Box<dyn Fn(&[&Data])->Data>
+}
+
+
+
+// =============
+// === Types ===
+// =============
+
+// === Types ===
+
+#[allow(missing_docs)]
+mod types {
+    use super::*;
+    pub type VarVec   = OptVec<Var,Index<Var>>;
+    pub type SheetVec = OptVec<Sheet,Index<Sheet>>;
+    pub type VarMap   = HashMapTree<String,Option<Index<Var>>>;
+    pub type SheetMap = HashMapTree<String,Index<Sheet>>;
+}
+use types::*;
+
+trait NewInstance<K> {
+    fn new_instance(&mut self) -> K;
+}
+
+impl NewInstance<Index<Var>> for VarVec {
+    fn new_instance(&mut self) -> Index<Var> {
+        self.insert_with_ix(|index| Var::new(index))
+    }
+}
+
+impl NewInstance<Index<Sheet>> for SheetVec {
+    fn new_instance(&mut self) -> Index<Sheet> {
+        self.insert_with_ix(|index| Sheet::new(index))
+    }
+}
+
+
+
+// ================
+// === Registry ===
+// ================
+
+/// Style sheet registry. Could be named "Cascading Style Sheets" but then the name will be
+/// confusing with CSS used in web development. Defines a set of cascading style sheets. Each
+/// style sheet can be assigned with a value of type `Data` or an expression to compute one. It
+/// also allows creating variables which are automatically bound to the most specific style sheet.
+/// See `Var` and `Sheet` to learn more.
+#[derive(Debug)]
+pub struct Registry {
+    /// Set of all variables.
+    pub vars : VarVec,
+    /// Set of all style sheets.
+    pub sheets : SheetVec,
+    /// Association of a path like 'button' -> 'size' to a variable.
+    pub var_map : VarMap,
+    /// Association of a path like 'button' -> 'size' to a style sheet.
+    pub sheet_map : SheetMap,
+}
+
+
+// === Constructors ===
+
+impl Registry {
+    /// Constructor.
+    pub fn new() -> Self {
+        let vars          = default();
+        let mut sheets    = OptVec::<Sheet,Index<Sheet>>::new();
+        let var_map       = default();
+        let root_sheet_id = sheets.new_instance();
+        let sheet_map     = SheetMap::from_value(root_sheet_id);
+        Self {vars,sheets,var_map,sheet_map}
+    }
+
+    /// Access variable by the given path or create new one if missing.
+    ///
+    /// Implementation note: under the hood, a `Sheet` for each sub-path will be created. For
+    /// example, when creating "panel.button.size" variable, three sheets will be created as well:
+    /// "panel.button.size", "button.size", and "size". This way we keep track of all possible
+    /// matches and we can create high-performance value binding algorithms.
+    pub fn var<P:Into<Path>>(&mut self, path:P) -> Index<Var> {
```

Comment:
I don't see how this method fill into `constructors` section :)

Reply:
Its a constructor of new variable. I'm leaving it heard as it matches my taste but as soon as you recommend something better Id be happy to move it! :)